sdk/src/main/java/com/bugsnag/android/Client.java

```diff
@@ -1237,16 +1250,21 @@ public void disableExceptionHandler() {
 
     void deliver(@NonNull Report report, @NonNull Error error) {
         try {
-            errorReportApiClient.postReport(config.getEndpoint(), report,
-                config.getErrorApiHeaders());
+            config.getDelivery().deliver(report, config);
             Logger.info("Sent 1 new error to Bugsnag");
-        } catch (NetworkException exception) {
-            Logger.info("Could not send error(s) to Bugsnag, saving to disk to send later");
-
-            // Save error to disk for later sending
-            errorStore.write(error);
-        } catch (BadResponseException exception) {
-            Logger.info("Bad response when sending data to Bugsnag");
+        } catch (DeliveryFailureException exception) {
+            switch (exception.reason) {
+                case CONNECTIVITY:
```

Comment:
Style nitpick: the Java convention is to use an exception hierarchy in these situations, any reason we are using an enum instead?

Reply:
@loopj it was suggested in the design phase that we use a single exception instead, rather than the previous `NetworkException` and `BadResponseException`. I think it'd make sense to continue using these as we were before with the new `Delivery` interface, as it would get rid of these switch statements and exception mapping code. @kattrali are you happy with this change?

Reply:
Specifically the implementation went in this direction as a response to [this comment](https://bugsnag.atlassian.net/browse/PLAT-1056?focusedCommentId=27929&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-27929):

> Should there be a generic `DeliveryFailureException` for outside use? Alternately, should the expected exception types be further documented and/or made generic?

@kattrali I'd be interested to know what the motiviation behind this was, or what your ideal implementation would like like?

I made the suggestion to @fractalwrench that there could be a hierarchy as follows:

```

Reply:
My thought was that we would had a single class for external implementors of `Delivery` to use to rethrow exceptions, not necessarily that we wouldn't throw a subclass of it internally and handle them differently. 

The intent was to reduce the concepts we need to explain for implementors, for example 

> if your `deliver` method fails, throw a `DeliveryFailureException` to retry

that way, they necessarily need to check different types of failures when rethrowing to trigger a later delivery.

Reply:
We would need to expose any subclasses of DeliveryFailureException so that the notifier can either save or discard the event.

Reply:
I prefer the second option as it feels more intuitive and will reduce the length of the catch block, but I can equally see that throwing a single exception type may make things easier to document.

Reply:
Exposing the subclasses is fine - the distinction just needs to be clear for implementors what will happen when throwing the different types so people know which will retry and which will be discarded.